File: androidApp/src/main/java/com/medhelp/medhelp/ui/profile/ProfilePresenter.java
package com.medhelp.medhelp.ui.profile;


import android.content.Context;
import com.google.gson.Gson;
import com.medhelp.medhelp.data.model.VisitResponse;
import com.medhelp.medhelp.data.network.NetworkManager;
import com.medhelp.medhelp.data.pref.PreferencesManager;
import com.medhelp.medhelp.ui.view.shopping_basket.sub.DataPaymentForRealm;
import com.medhelp.medhelp.utils.timber_log.LoggingTree;
import com.medhelp.shared.model.UserResponse;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class ProfilePresenter implements ProfileHelper.Presenter {

    Context context;
    ProfileHelper.View viewHelper;

    PreferencesManager prefManager;
    NetworkManager networkManager;

    public ProfilePresenter(Context context, ProfileHelper.View viewHelper) {
        this.context = context;

        this.viewHelper = viewHelper;

        prefManager = new PreferencesManager(context);
        networkManager = new NetworkManager(prefManager);
    }


    @Override
    public void getVisits() {
        final String[] time = new String[1];
        final String[] today = new String[1];

        viewHelper.showLoading();
        CompositeDisposable cd = new CompositeDisposable();
        cd.add(networkManager
                .getCurrentDateApiCall()
                .concatMap(resp -> {
                    time[0] = resp.getResponse().getTime();
                    today[0] = resp.getResponse().getToday();
                    return networkManager.getAllReceptionApiCall();
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response ->
                        {
                            if (viewHelper == null)
                                return;

                            if (!response.isError())
                                viewHelper.updateData(response.getResponse(), today[0], time[0]);
                            else
                                viewHelper.showErrorScreen();

                            viewHelper.hideLoading();
                            viewHelper.swipeDismiss();
                            cd.dispose();
                        },
                        throwable -> {
                            Timber.tag("my").e(LoggingTree.getMessageForError(throwable, "ProfilePresenter$getVisits "));

                            if (viewHelper == null)
                                return;

                            viewHelper.hideLoading();
                            viewHelper.swipeDismiss();
                            viewHelper.showErrorScreen();
                            cd.dispose();
                        }));
    }

    @Override
    public void updateHeaderInfo() {
        viewHelper.updateHeader(prefManager.getCenterInfo());
    }

    @Override
    public void unSubscribe() {
        viewHelper.hideLoading();
    }

    @Override
    public void cancellationOfVisit(int idUser, int id_record, String cause, int idBranch) {
        viewHelper.showLoading();

        new CompositeDisposable().add(networkManager
                .sendCancellationOfVisit(idUser, id_record, cause, idBranch)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response ->
                        {
                            Timber.tag("my").v("Отмена приема id записи " + id_record + " причина " + cause);
                            getVisits();
                        }
                        , throwable ->
                        {
                            viewHelper.hideLoading();
                            Timber.tag("my").e(LoggingTree.getMessageForError(throwable, "ProfilePresenter$cancellationOfVisit "));
                            viewHelper.showError("Произошла ошибка при выполнении операции, попробуйте повторить");
                        }
                ));
    }

    @Override
    public void confirmationOfVisit(int idUser, int id_record, int idBranch) {
        viewHelper.showLoading();

        new CompositeDisposable().add(networkManager
                .sendConfirmationOfVisit(idUser, id_record, idBranch)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response ->
                        {
                            Timber.tag("my").v("Подтвердение приема id записи %s", id_record);
                            if (response.getResponse()) {
                                getVisits();
                            }
                        }
                        , throwable ->
                        {
                            Timber.tag("my").e(LoggingTree.getMessageForError(throwable, "ProfilePresenter$confirmationOfVisit "));
                            viewHelper.showError("Произошла ошибка при выполнении операции, попробуйте повторить");
                        }
                )
        );

        viewHelper.hideLoading();
    }

    @Override
    public String getCurrentHospitalBranch() {
        //String ss=getDataHelper().getCurrentNameBranch();
        return prefManager.getCurrentUserInfo().getNameBranch();
    }

    @Override
    public String getUserToken() {
        return prefManager.getCurrentUserInfo().getApiKey();
    }


    @Override
    public UserResponse getCurrentUser() {
        return prefManager.getCurrentUserInfo();
    }

    @Override
    public void setCurrentUser(UserResponse user) {
        prefManager.setCurrentUserInfo(user);
    }


    private int getCountItems(String element) {
        return element.length() - element.replace("&", "").length();
    }

    private void sendToServer(DataPaymentForRealm data) {
        Timber.tag("my").v("услуги оплачены, Sum: " + data.getPrice() + "; IdRecord: " + data.getIdZapisi() + "; IdPayment: " + data.getIdPayment() + "; IdYsl: " + data.getIdYsl());

        CompositeDisposable cd = new CompositeDisposable();

        cd.add(networkManager
                .sendToServerPaymentData(data.getIdUser(), data.getIdBranch(), data.getIdZapisi(), data.getIdYsl(), data.getPrice(), getCountItems(data.getIdUser()), data.getIdPayment())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        response -> {
                            if (response.getResponse()) {

                            } else {
                                Gson gson = new Gson();
                                Timber.tag("my").e(LoggingTree.getMessageForError(null, "ProfilePresenter/sendPaymentToServer, совпал pay_id " + gson.toJson(data)));
                            }

                            prefManager.deletePaymentData(data);

                            cd.dispose();

                        }, throwable -> {
                            Gson gson = new Gson();
                            Timber.tag("my").e(LoggingTree.getMessageForError(throwable, "ProfilePresenter/sendPaymentToServer : " + gson.toJson(data)));
                            cd.dispose();
                        }
                )
        );
    }

    @Override
    public void sendConfirmComing(VisitResponse viz) {

        viewHelper.showLoading();
        CompositeDisposable cd = new CompositeDisposable();
        cd.add(networkManager
                .sendIAmHere(viz.getIdUser(), viz.getIdRecord(), viz.getIdBranch())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        response -> {

                            viewHelper.hideLoading();
                            viewHelper.responseConfirmComing(viz);
                            cd.dispose();
                        }, throwable -> {
                            viewHelper.hideLoading();
                            Timber.tag("my").e(LoggingTree.getMessageForError(throwable, "ProfilePresenter/sendConfirmComing "));
                            cd.dispose();
                        }
                )
        );
    }
}

File: androidApp/src/main/java/com/medhelp/medhelp/ui/profile/ProfilePresenter.kt
package com.medhelp.medhelp.ui.profile

import android.content.Context
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.medhelp.medhelp.data.model.DataPaymentForRealm
import com.medhelp.medhelp.data.pref.PreferencesManager
import com.medhelp.newmedhelp.model.VisitResponseAndroid
import com.medhelp.medhelp.utils.timber_log.LoggingTree.Companion.getMessageForError
import com.medhelp.medhelp.utils.TimesUtils
import com.medhelp.newmedhelp.model.VisitResponse
import com.medhelp.shared.model.UserResponse
import com.medhelp.shared.network.NetworkManager
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import timber.log.Timber
import java.util.ArrayList

class ProfilePresenter(val context: Context, val viewHelper: ProfileFragment){
    var prefManager: PreferencesManager
    var networkManager: NetworkManager

    val mainScope = MainScope()

    init {
        prefManager = PreferencesManager(context)
        networkManager = NetworkManager()
    }

    fun getVisits1() {
        viewHelper!!.showLoading()
        mainScope.launch {
            kotlin.runCatching {
                networkManager.getCurrentDateApiCall(prefManager.currentUserInfo!!.apiKey!!, prefManager.centerInfo!!.db_name!!,prefManager.currentUserInfo!!.idUser.toString(),prefManager.currentUserInfo!!.idBranch.toString())
            }
                .onSuccess {
                    getVisits2(it.response!!.time!!, it.response!!.today!!)
                }.onFailure {
                    Timber.tag("my").e(getMessageForError(it, "ProfilePresenter\$getVisits1 "))
                    if (viewHelper == null) return@onFailure
                    viewHelper!!.hideLoading()
                    viewHelper!!.swipeDismiss()
                    viewHelper!!.showErrorScreen()
                }
        }
    }

    fun getVisits2(time : String, today : String){

        mainScope.launch {
            kotlin.runCatching {
                networkManager.getAllReceptionApiCall(prefManager.currentUserInfo!!.apiKey!!, prefManager.centerInfo!!.db_name!!,prefManager.currentUserInfo!!.idUser.toString(),prefManager.currentUserInfo!!.idBranch.toString())
            }
                .onSuccess {
                    if (viewHelper == null) return@onSuccess

                    if (!it.error) {
                        var listAndr = mutableListOf<VisitResponseAndroid>()
                        for(i in it.response){
                            listAndr.add(VisitResponseAndroid(i))
                        }

                        viewHelper!!.updateData(listAndr, today, time)
                    }else
                        viewHelper!!.showErrorScreen()
                    viewHelper!!.hideLoading()
                    viewHelper!!.swipeDismiss()
                }.onFailure {
                    Timber.tag("my").e(getMessageForError(it, "ProfilePresenter\$getVisits2 "))
                    if (viewHelper == null) return@onFailure
                    viewHelper!!.hideLoading()
                    viewHelper!!.swipeDismiss()
                    viewHelper!!.showErrorScreen()
                }
        }
    }

    fun updateHeaderInfo() {
        prefManager.centerInfo?.let {
            viewHelper!!.updateHeader(it)
        }
    }

    fun unSubscribe() {
        viewHelper!!.hideLoading()
    }

    fun cancellationOfVisit(idUser: Int, id_record: Int, cause: String, idBranch: Int) {
        viewHelper!!.showLoading()

        mainScope.launch {
            kotlin.runCatching {
                networkManager.sendCancellationOfVisit(
                    idUser.toString(),
                    id_record.toString(),
                    cause,
                    TimesUtils.getCurrentDate(TimesUtils.DATE_FORMAT_ddMMyy),
                    idBranch.toString(),
                    prefManager.currentUserInfo!!.apiKey!!,
                    prefManager.centerInfo!!.db_name!!,
                    prefManager.currentUserInfo!!.idUser.toString(),
                    prefManager.currentUserInfo!!.idBranch.toString()
                )
            }
                .onSuccess {
                    Timber.tag("my").v("Отмена приема id записи $id_record причина $cause")
                    getVisits1()
                }.onFailure {
                    viewHelper!!.hideLoading()
                    Timber.tag("my")
                        .e(getMessageForError(it, "ProfilePresenter\$cancellationOfVisit "))
                    viewHelper!!.showError("Произошла ошибка при выполнении операции, попробуйте повторить")
                }
        }
    }

    fun confirmationOfVisit(idUser: Int, id_record: Int, idBranch: Int) {
        viewHelper!!.showLoading()

        mainScope.launch {
            kotlin.runCatching {
                networkManager.sendConfirmationOfVisit(
                    idUser.toString(),
                    id_record.toString(),
                    idBranch.toString(),
                    prefManager.currentUserInfo!!.apiKey!!,
                    prefManager.centerInfo!!.db_name!!,
                    prefManager.currentUserInfo!!.idUser.toString(),
                    prefManager.currentUserInfo!!.idBranch.toString()
                )
            }
                .onSuccess {
                    Timber.tag("my").v("Подтвердение приема id записи %s", id_record)
                    if (it.response) {
                        getVisits1()
                    }
                }.onFailure {
                    Timber.tag("my")
                        .e(getMessageForError(it, "ProfilePresenter\$confirmationOfVisit "))
                    viewHelper!!.showError("Произошла ошибка при выполнении операции, попробуйте повторить")
                    viewHelper!!.hideLoading()
                }
        }
    }

    fun getCurrentHospitalBranch(): String {
        //String ss=getDataHelper().getCurrentNameBranch();
        return prefManager.currentUserInfo!!.nameBranch!!
    }

    fun getUserToken(): String {
        return prefManager.currentUserInfo!!.apiKey!!
    }

    fun getCurrentUser(): UserResponse {
        return prefManager.currentUserInfo!!
    }

    fun setCurrentUser(user: UserResponse) {
        prefManager.currentUserInfo = user
    }

    fun getCountItems(element: String): Int {
        return element.length - element.replace("&", "").length
    }

    fun sendToServer(data: DataPaymentForRealm) {
        Timber.tag("my").v("услуги оплачены, Sum: " + data.price + "; IdRecord: " + data.idZapisi + "; IdPayment: " + data.idPayment + "; IdYsl: " + data.idYsl)

        mainScope.launch {
            kotlin.runCatching {
                networkManager.sendToServerPaymentData(data.idUser, data.idBranch, data.idZapisi, data.idYsl, data.price, getCountItems(data.idUser).toString(), data.idPayment,
                    prefManager.currentUserInfo!!.apiKey!!, prefManager.centerInfo!!.db_name!!,prefManager.currentUserInfo!!.idUser.toString(),prefManager.currentUserInfo!!.idBranch.toString())
            }
                .onSuccess {
                    if (it.response=="true") {
                    } else {
                        val gson = Gson()
                        Timber.tag("my").e(getMessageForError(null, "ProfilePresenter/sendPaymentToServer, совпал pay_id " + gson.toJson(data)))
                    }
                    prefManager.deletePaymentData(data)
                }.onFailure {
                    val gson = Gson()
                    Timber.tag("my").e(getMessageForError(it, "ProfilePresenter/sendPaymentToServer : " + gson.toJson(data)))
                }
        }
    }

    fun sendConfirmComing(viz: VisitResponseAndroid) {
        viewHelper!!.showLoading()

        mainScope.launch {
            kotlin.runCatching {
                networkManager.sendIAmHere(viz.idUser.toString(), viz.idRecord.toString(), viz.idBranch.toString(),
                    prefManager.currentUserInfo!!.apiKey!!, prefManager.centerInfo!!.db_name!!,prefManager.currentUserInfo!!.idUser.toString(),prefManager.currentUserInfo!!.idBranch.toString())
            }
                .onSuccess {
                    viewHelper!!.hideLoading()
                    viewHelper!!.responseConfirmComing(viz)
                }.onFailure {
                    viewHelper!!.hideLoading()
                    Timber.tag("my").e(getMessageForError(it, "ProfilePresenter/sendConfirmComing "))
                }
        }
    }
}